src/translator/service.cpp

```diff
@@ -135,8 +158,43 @@ AsyncService::AsyncService(const AsyncService::Config &config)
       safeBatchingPool_(),
       cache_(makeOptionalCache(config_.cacheSize, /*mutexBuckets=*/config_.numWorkers)),
       logger_(config.logger) {
-  ABORT_IF(config_.numWorkers == 0, "Number of workers should be at least 1 in a threaded workflow");
-  workers_.reserve(config_.numWorkers);
+  if (config_.gpuWorkers.empty()) {
+    ABORT_IF(config_.numWorkers == 0, "Number of workers should be at least 1 in a threaded workflow");
+    workers_.reserve(config_.numWorkers);
+  } else {
+    if (config_.numWorkers != 0) LOG(info, "Unable to mix GPU and CPU workers, using GPU workers only...");
+    workers_.reserve(config_.gpuWorkers.size());
+    // VERY VERY HACKY. EVERYTHING USES NUM_WORKERS AS A REFERENCE FOR THE NUMBER OF WORKERS,
+    // REFACTOR TO USE gpuWorkers directly...
+    config_.numWorkers = config_.gpuWorkers.size();
+  }
+  // Initiate terminology map if present
+  if (!config_.terminologyFile.empty()) {
+    // Create an input filestream
+    std::ifstream myFile(config_.terminologyFile);
+
+    // Make sure the file is open
+    if (!myFile.is_open()) throw std::runtime_error("Could not open file: " + config_.terminologyFile);
+    std::string line;
+    std::unordered_map<std::string, std::string> tempmap;  // Read in the TSV to here
+    while (std::getline(myFile, line)) {
+      // Create a stringstream of the current line
+      std::stringstream ss(line);
+
+      std::string srcword;
+      std::string replacementword;
+      getline(ss, srcword, '\t');
+      getline(ss, replacementword, '\n');  // BEWARE of windows file ndings
```

Comment:
This is why I would have liked it if we could have kept the "read the tsv file" part out of the bergamot C++ codebase, and just implement it in Python/Qt/JavaScript which are much better at integrating with the local environment.

… But treat this complaint so I'll stop complaining about this from now on!

Reply:
Well, functionally, we can bypass the redundant terminology implementation, by applying terminology constraints directly. Maybe we can add this to the readme. I would like to keep a C++ implementation at the very least as documentation.